### ISSUE TITLE
Implement role-based access control

### ISSUE THREAD

**Issue by stwalkerster** [enwikipedia-acc/waca#283]
Allow granular permissions to be defined and assigned to roles.

Security will be implemented two-fold, firstly the User and Admin values in the status field on users will now be used for determining if a user is currently active or not. These checks will be done before applying roles to the current user.

The next stage is every page will implicitly require permissions at an action level. These permissions will mostly be defined implicitly, such as `PageWelcomeTemplate/create`, and as such don't need to be declared or tested by the page.

`InternalPageBase` will still handle the actual security barrier and displaying this to the user, but only in the capacity of implementing the decision. `SecurityManager` will be responsible for making the decision.

Identification checking will be done on a role-by-role basis. If a role requires (or rather, is not exempt) from identification checks, it won't be considered by `SecurityManager` in the list of rights granted to a user, unless the check is a rejection, in which case it will be considered to see if the user would have had that right to choose a more correct error condition (not allowed vs not identified)

## User changes

* [x] The user status column values will be replaced with "Active". 
* [x] The checkuser column will also be dropped.

## Roles

Roles are to be defined in a single place, probably a static class called `RoleConfiguration`.

* [x] Create three roles - user, admin and checkuser. Assign rights approximately as they are defined currently, skipping duplicated rights as needed, such that a checkuser access is distinct from tool admin, but such that they can perform the tasks they need to
* [x] Role allocation to users will be done by a new database table storing a unique set of user IDs and role names. Table to be managed using the standard `DataObject` infrastructure.
* [x] Migrate all users to relevant roles.

![role](https://cloud.githubusercontent.com/assets/722710/22130934/5da2afdc-dea7-11e6-9612-b7b4a9d2e9f6.png)

## Page access control

* [x] Kill off `getSecurityConfiguration()`.
* [x] Modify `InternalPageBase` to pass authorisation requests to a helper class.
* [x] Create some way for pages to explicitly skip all access checks (so you don't get access denied errors because you're not logged in while loading the login form...). This is to only be used for "critical" pages that break the tool if they're forbidden at any time, such as the login form.
* [x] Implement some explicit rights check barrier that pages can use for determining user access, similar to `barrierTest` in cases such as displaying an edit button vs a view button
* [x] Kill off `barrierTest()` accesses from templates. This was a bad idea when I wrote it.

## Security manager

Rough process will be:
1) Request received by security manager for rights check
2) Security manager requests roles for user.
3) Check identification status of user
4) Split roles user has by `($roleRequiresIdentification && $userIdentified || !$roleRequiresIdentification)`
5) Cache steps 2 through 4.
5) Take first pile of roles and check the requested right exists in the set. If so, return success.
6) Take the second pile of roles and check the requested right exists in the set. If so, return ERR_NOT_IDENTIFIED.
7) Return general failure.

* [x] Implement it
* [ ] Unit test the hell out of this.
* [ ] Caching (check performance when not remote)

## Registration

* [x] Generalise the registration form, so different instances can assign users to a different initial role.
* [x] Figure out the UX for getting to the right signup form. Perhaps a pre-signup step along the lines of a wizard?

## User management
* [x] The standard promote/demote workflow actions are now not applicable. We need a way of granting/removing roles from users. Special:GroupPermissions is probably going to be the inspiration for this.
* [x] Think of a way to stop tool admins granting themselves checkuser access. Probably something similar to `$wgAddGroups` defined in `RoleConfiguration`
* [x] Adapt the user status buttons as applicable - probably won't need many changes.

## Misc
* [x] Update documentation
* [x] Deprecate `User::isCheckuser()`
* [x] Deprecate `User::isAdmin()`

**Comment by FastLizard4** [comment]
Alright, so here are my thoughts looking at this:

* For the new userrole table, why not get rid of the `id` column and just do a composite primary key on `user` and `role`?
* "Create some way for pages to explicitly skip all access checks (so you don't get access denied errors because you're not logged in while loading the login form...)" this sounds nocheck-tastic.  :P Perhaps a better approach would be to allow permissions (such as a hypothetical "login" permission, in this case) to be granted to an anonymous role?
* I'm wary of caching any aspect of the access control system beyond identified status as already written, and that's really only so we are more resistant to Meta-Wiki API problems/slowness.  Caching when it comes access control is always a bottomless abyss of trouble, and I think it's better to do without caching unless absolutely necessary.
* For user registration, I think the cleanest solution from the end-user perspective is to have separate registration forms; this also allows us to present different relevant information depending on the type of ACC account being sought

Otherwise, everything looks solid to me.

**Comment by stwalkerster** [comment]
1) For the userrole table, I was looking to use this via the standard DataObject infrastructure for managing the roles a user has at an underlying level. DataObject requires two columns to be present on every table, `id` and `updateversion`. The alternative is to manage this table manually - given the simplicity of the table I doubt the added complexity in doing so would be too much work. It just happens outside the method pretty much all the other database access occurs - that was my reason for preferring DataObject.
2) "nocheck-tastic"! Indeed, when you put it that way, it does seem a little like that. However, my reason for stating "explicit" was more along the lines of a page class being forced to implement a method along these lines:
```php
<?php
class InternalPageBase extends PageBase
{
    // other stuff...

    // Base implementation defaults to protected
    protected function isProtectedPage() {
        return true;
    }

    // other stuff...
}

class PageDoStuff extends InternalPageBase
{
    // other stuff...

    // subclass explicitly overrides the access control checks
    protected function isProtectedPage() {
        return false;
    }

    // other stuff...
}
```

I was thinking about an anonymous role too, and indeed was thinking of implementing something like that. However, certain pages, such as PageLogin, are a little more special technically than any other page, and these pages (logout, register, 404, there's a couple of others too) make no sense ever being part of an access group, but should be accessible to everyone all the time. That's what the explicit skip of access checks is intended for, not for making a page accessible to the public - drawing a line between "ooh, someone might want to see PageTeam without being logged in" and "the tool cannot technically function unless this page is accessible without being logged in".

3) The caching isn't intended to be cross-request, but rather for when multiple right-checks are needed within a single request. If user rights change mid-request and we need to respect that, then we need a rethink about a lot of assumptions made in the tool. Note that even if we do cache user rights for more than a single request, I'd not be wanting to cache user status checks too - under this new structure, user rights aren't revoked on a user status change, but are instead distinct, thus allowing a user to be both suspended and a tool admin - this is intentional to distinguish between normal users and event users. How this is managed socially is `not-my-problem.gif` 😛.

4) Sure, requires a little rethink on my part as to how this is designed, but should be possible along the same lines as custom request forms for event management.

**Comment by FastLizard4** [comment]
I'll go numeric with my replies too.  :P

1. Okay, that's fair enough; I just saw that column and thought "unnecessary overhead" there.  I don't think it'll be a significant performance difference either way, so let's go for consistency.
2. Ahh, okay, I see what you mean now.  I'd still advocate for an anonymous role for the "normal" pages we want to make public, but on the "public page" vs "public visibility required to function" division, the approach you give for the latter looks good.
3. Oooohhhhhhhh, okay, this is just me misunderstanding what you were getting at then.  Yeah, I think it's very reasonable to assume that within a single request that a user's rights will not change.
4. *This space intentionally left blank.*

:+1:

**Comment by stwalkerster** [comment]
I've updated the spec with a) the comments above, and b) further points noticed during development, such as access control that has been implemented by directly checking `User::isCheckuser()` currently.

User::isCheckuser() is going to be somewhat meaningless after this, so I'm going to remove it, and we can let the roles define the relevant permissions. We can still build reports and such based on the roles, just from a code perspective it shouldn't make any difference if a user is in role `foo` and can do `bar`, or if they are in role `garply` and can do `splurk`. They either can do `splurk` or they can't.
